Auto update

GabeIsman opened the pull request:
This adds auto update functionality.

### On OSX
It leverages the [Squirrel framework](https://github.com/Squirrel/Squirrel.Mac) via electron's built-in auto-update library for a pretty seamless experience. On startup it checks for updates by pinging a [nuts server](https://github.com/GitbookIO/nuts), which is uses github releases as the only backend. If an update is found it is downloaded in the background and the user is prompted to install it, which restarts the app. The user can decline to install the update.

### On Linux
It pings the public github releases API and compares the tag name of the latest release to the version that is currently running. If the current version string is not a substring of the latest then the user is notified that there is an update that should be installed through their package manager, or by visiting the release page (which is included in the message, although not linked). This version comparison is a bit of a hack, since it will result in a false negative when comparing `0.0.1` and `v0.0.12` for example. We could easily improve this by using [semver](https://github.com/npm/node-semver) or similar. This does require that we be careful about our tag names, which must be well-formatted version strings.

### Windows
Theoretically this functionality should just work on Windows, but I haven't tested it since we aren't currently supporting Windows.

GabeIsman committed:
WIP

GabeIsman committed:
Merge branch 'master' into auto-update

GabeIsman committed:
Light cleanup

GabeIsman committed:
Auto update fixes.

GabeIsman committed:
Fix quit and install hanging

Workaround discovered here:
https://github.com/electron/electron/issues/3583

GabeIsman committed:
Clean up development logging, urls.

GabeIsman committed:
Use tag_name instead of name.

garrettr reviewed (changes requested):
@GabeIsman Thanks for your work on this, and sorry for the long delay in reviewing it! I am going to go ahead and add a few commits to this PR to address most of my comments, so don't worry about making any code changes for this review; however, there are a few questions I've asked that I would appreciate your experience in answering!

Also, @conorsch explained to me that the Squirrel vs. App Store approaches for auto-updates on Mac OS X are mutually exclusive, and I guess we haven't made a final call about which one we want to use for the initial release. I think we should merge this Squirrel code for now, and revisit this issue as we continue driving towards the initial release.

garrettr committed:
Minor review fixes

garrettr committed:
Merge branch 'master' into auto-update

garrettr committed:
Use semver to compare versions

Adds the semver library to perform the comparison. If either the "latest
release" version fetched from GitHub or the currentVersion of the
running app is invalid according to the semver library, an error message
is logged to the console, but no message is displayed to the end user.

This behavior should allow us to recover from any mistakes in setting
the latest release tag seamlessly and without needlessly alarming end
users.

YiiKuoChong commented:
Do you plan to auto-update through TOR or a proxy?

conorsch commented:
See #160. We're no longer actively maintaining Sunder, so this PR will not be merged. Closing.
